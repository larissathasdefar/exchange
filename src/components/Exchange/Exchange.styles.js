import styled from 'styled-components'
import Paper from '@material-ui/core/Paper'
import TextField from '@material-ui/core/TextField'
import Button from '@material-ui/core/Button'
import SwapIcon from '@material-ui/icons/SwapVert'

export const Container = styled(Paper)`
  max-width: calc(600px - 40px);
  padding: 20px 14px;
  width: 100%;
  margin-bottom: 20px;
`

export const CurrencyContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: start;
`

export const Currency = styled.div`
  margin-right: ${(({ margin }) => margin)}px;
`

export const TextInput = styled(TextField)`
  width: 120px;
  float: right;
  max-height: 68px;
`

export const Complement = styled.div`
  padding-left: 40px;
  min-width: 150px;
`

export const LoadingContainer = styled(Complement)`
  & div {
    height: 67px;
    align-items: baseline;
  }
`

export const StepContainer = styled.div`
  width: fit-content;
  margin: 0 auto;
`

export const ButtonsContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`

export const Cancel = styled(Button)`
  margin-right: 12px !important;
`

export const ExchangeIcon = styled(SwapIcon)`
  && {
    margin: 0 auto;
    display: block;
    font-size: 32px;
    color: #438ac3;
  }
`
