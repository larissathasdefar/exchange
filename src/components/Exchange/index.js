import React, { PureComponent, Fragment } from 'react'
import _ from 'prop-types'
import Typography from '@material-ui/core/Typography'
import Button from '@material-ui/core/Button'
import InputAdornment from '@material-ui/core/InputAdornment'
import MoneyIcon from '@material-ui/icons/AttachMoney'
import Stepper from '../UI/Stepper'
import {
  Container,
  Currency,
  CurrencyContainer,
  TextInput,
  Complement,
  StepContainer,
  ButtonsContainer,
  Cancel,
  ExchangeIcon,
} from './Exchange.styles'

class Exchange extends PureComponent {
  state={
    convert: '',
    openConfirmation: false,
  }

  handleToggleConfirmation = () => {
    this.setState(({ openConfirmation }) => ({ openConfirmation: !openConfirmation }))
  }

  handleConfirm = () => {
    this.handleToggleConfirmation()
    // TODO: Call Api
  }

  handleChange = event => {
    const { value } = event.target
    if (parseFloat(value) || value === '') {
      this.setState({ convert: value })
    }
  }

  renderField = max => {
    const { convert } = this.state
    return (
      <Complement>
        <TextInput
          autoFocus
          label="Convert"
          value={convert}
          max={max}
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                { convert === '' ? '' : '-' }
              </InputAdornment>
            ),
          }}
          InputLabelProps={{
            shrink: true,
          }}
          helperText="X1 = X0.44"
          onChange={this.handleChange}
        />
      </Complement>
    )
  }

  renderConversion = () => {
    const { convert } = this.state
    const parsed = parseFloat(convert) || 0
    return (
      <Complement>
        <Typography variant="h4" align="right">
          { `+ ${parsed.toFixed(2)}` }
        </Typography>
        <Typography variant="subtitle1" align="right">
          X1 = X0.44
        </Typography>
      </Complement>
    )
  }

  renderCurrency = ({ title, subtitle, from, max }) => {
    const { convert } = this.state
    return (
      <CurrencyContainer>
        <Currency>
          <Typography variant="h4" align="center">
            { title }
          </Typography>
          <Typography variant="subtitle1" align="center">
            { subtitle }
          </Typography>
        </Currency>
        { from ? this.renderField(max) : this.renderConversion() }
      </CurrencyContainer>
    )
  }

  render() {
    return (
      <Container elevation={1}>
        <Typography variant="h4">Let's Exchange</Typography>
        <Typography variant="caption" paragraph>How much would you like to exchange?</Typography>
        <StepContainer>
          <Stepper
            steps={[
              { from: true, title: 'GPB', subtitle: 'You have £48.54', max: 48.54 },
              { from: true, title: 'EUR', subtitle: 'You have €138.40', max: 138.40 },
              { from: true, title: 'USD', subtitle: 'You have $60.00', max: 60.00 },
            ]}
            renderStep={this.renderCurrency}
          />
        </StepContainer>
        <ExchangeIcon />
        <StepContainer>
          <Stepper
            steps={[
              { to: true, title: 'GPB', subtitle: 'You have £48.54', max: 48.54 },
              { to: true, title: 'EUR', subtitle: 'You have €138.40', max: 138.40 },
              { to: true, title: 'USD', subtitle: 'You have $60.00', max: 60.00 },
            ]}
            start={1}
            renderStep={this.renderCurrency}
          />
        </StepContainer>
        <ButtonsContainer>
          <Cancel variant="outlined" color="secondary">Cancel</Cancel>
          <Button variant="contained" color="primary">
            Exchange
            <MoneyIcon />
          </Button>
        </ButtonsContainer>
      </Container>
    )
  }
}

export default Exchange
