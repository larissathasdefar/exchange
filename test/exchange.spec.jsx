import React from 'react'
import { shallow } from 'enzyme'

import Exchange from 'components/Exchange'

describe('Exchange', () => {
  it('Having 2 pockets', () => {
    const component = shallow(
      <Exchange
        user={{
          pockets: [
            { code: 'USD', amount: 44.90 },
            { code: 'GBP', amount: 50.56 },
            { code: 'EUR', amount: 90.06 },
          ],
          photo: 'https://scontent.fjoi1-2.fna.fbcdn.net/v/t1.0-1/c21.0.699.699a/s100x100/13240613_1031098473634742_623670164829509408_n.jpg?_nc_cat=106&_nc_ht=scontent.fjoi1-2.fna&oh=73ef6448c5749f90458e20e6bd8afe64&oe=5CEC15C9',
          name: 'Larissa Farias',
          id: 1,
          followedRates: [
            { from: 'USD', to: 'GBP' },
            { from: 'EUR', to: 'GBP' },
            { from: 'GBP', to: 'USD' },
          ],
        }}
        currencies={{
          USD: 'United States Dollar',
          GBP: 'British Pound Sterling',
          EUR: 'Euro',
        }}
        history={{ push: route => console.log(route) }}
        rates={{
          USD: 1.129452,
          GBP: 0.882712,
          EUR: 1
        }}
        onLoadRates={jest.fn()}
        onConfirmExchange={jest.fn()}
      />
    )
    expect(component.debug()).toMatchSnapshot()
  })
})