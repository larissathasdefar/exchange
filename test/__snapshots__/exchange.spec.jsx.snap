// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Having 2 pockets 1`] = `
"<Styled(WithStyles(Paper)) elevation={1}>
  <WithStyles(Typography) variant=\\"h4\\" align=\\"center\\">
    Let&#39;s Exchange
  </WithStyles(Typography)>
  <WithStyles(Typography) variant=\\"caption\\" align=\\"center\\" paragraph={true}>
    How much would you like to exchange?
  </WithStyles(Typography)>
  <styled.div>
    <CustomStepper steps={{...}} start={0} renderStep={[Function]} onChangeStep={[Function]} align=\\"inherit\\" />
  </styled.div>
  <Styled(pure(SwapVertIcon)) />
  <styled.div>
    <CustomStepper steps={{...}} start={1} renderStep={[Function]} onChangeStep={[Function]} align=\\"inherit\\" />
  </styled.div>
  <styled.div>
    <Styled(WithStyles(Button)) variant=\\"outlined\\" color=\\"secondary\\" onClick={[Function: onClick]}>
      Cancel
    </Styled(WithStyles(Button))>
    <WithStyles(Button) variant=\\"contained\\" color=\\"primary\\" disabled={false} onClick={[Function]}>
      Exchange
      <pure(AttachMoneyIcon) />
    </WithStyles(Button)>
  </styled.div>
  <WithStyles(Dialog) open={false} onClose={[Function]}>
    <WithStyles(DialogTitle)>
      Error
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)>
      <WithStyles(DialogContentText)>
        You must fill the field with a number.
      </WithStyles(DialogContentText)>
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <Styled(WithStyles(Button)) variant=\\"contained\\" color=\\"primary\\" onClick={[Function]}>
        Close
      </Styled(WithStyles(Button))>
    </WithStyles(DialogActions)>
  </WithStyles(Dialog)>
</Styled(WithStyles(Paper))>"
`;
